src/ethereum/osaka/bal_tracker.py

```diff
@@ -0,0 +1,118 @@
+
+
+class StateChangeTracker:
+    """
+    Tracks state changes during transaction execution for BAL construction.
+    """
+
+    def __init__(self, bal_builder: BALBuilder):
+        self.bal_builder = bal_builder
+        self.pre_storage_cache: Dict[tuple, U256] = {}  # (address, key) -> pre_value
+        self.current_tx_index: int = 0
+
+    def set_transaction_index(self, tx_index: int) -> None:
+        """Set the current transaction index for tracking changes."""
+        self.current_tx_index = tx_index
+        # Note: We keep the pre_storage_cache across transactions within the same block
+        # as we need it to determine what was the original state before the block
+    
+    def capture_pre_state(self, address: Address, key: Bytes, state: State) -> U256:
+        """Capture and cache the pre-state value for a storage location."""
+        cache_key = (address, key)
+        if cache_key not in self.pre_storage_cache:
+            self.pre_storage_cache[cache_key] = get_storage(state, address, key)
+        return self.pre_storage_cache[cache_key]
+
+    def track_address_access(self, address: Address) -> None:
+        """Track that an address was accessed (even if not changed)."""
+        self.bal_builder.add_touched_account(address)
+
+    def track_storage_read(self, address: Address, key: Bytes, state: State) -> None:
+        """Track a storage read operation."""
+        self.track_address_access(address)
+        
+        # Capture pre-state value for potential later comparison
+        self.capture_pre_state(address, key, state)
+        
+        # Add as read (will be filtered out later if this slot is written to)
+        self.bal_builder.add_storage_read(address, key)
+
+    def track_storage_write(
+        self, 
+        address: Address, 
+        key: Bytes, 
+        new_value: U256, 
+        state: State
+    ) -> None:
+        """Track a storage write operation."""
+        self.track_address_access(address)
+        
+        # Get pre-value to determine if this is actually a change
+        pre_value = self.capture_pre_state(address, key, state)
+        
+        # Convert U256 to 32-byte value
+        value_bytes = new_value.to_be_bytes32()
+        
+        # Only track as write if value actually changed
+        if pre_value != new_value:
+            self.bal_builder.add_storage_write(address, key, self.current_tx_index, value_bytes)
+        else:
+            # Unchanged write - track as read instead
+            self.bal_builder.add_storage_read(address, key)
+
+    def track_balance_change(
+        self, 
+        address: Address, 
+        new_balance: U256, 
+        state: State
+    ) -> None:
+        """Track a balance change."""
+        self.track_address_access(address)
+        
+        # Convert U256 to 16-byte balance (uint128 - sufficient for total ETH supply)
+        balance_bytes = new_balance.to_be_bytes32()[-16:]  # Take last 16 bytes for uint128
```

Comment:
This should probably be:

```suggestion
        balance_bytes = U128(new_balance).to_be_bytes16()
```

In the unlikely event that `new_balance` is ever greater than `U128.MAX_VALUE`, we'd want to throw an exception instead of silently truncating.

This might require a change and new release of ethereum-types.